packages/connect/src/base/listener.ts

```diff
@@ -377,24 +184,162 @@ class Listener extends EventEmitter implements InterfaceListener {
+      tcpSocket.destroy()
+      secondUdpSocket.close()
+    }
+
+    const tcpSocket = createConnection({
+      port,
+      host: externalIp,
+      localPort: 9092,
+      signal: abort.signal
+    })
+      .on('connect', () => {
+        tcpSocket.write(TCP_TEST, (err: any) => {
+          if (err) {
+            log(`Failed to send TCP packet`, err)
+          }
+        })
       })
+      .on('error', (err: any) => {
+        if (err && (err.code == undefined || err.code !== 'ABORT_ERR')) {
+          error(`Error while checking NAT situation`, err.message)
+        }
+      })
+
+    if (!done) {
+      const results = await Promise.allSettled([waitForIncomingUdpPacket.promise, waitForIncomingTcpPacket.promise])
+
+      cleanUp()
+
+      return {
+        udpMapped: results[0].status === 'fulfilled',
+        tcpMapped: results[1].status === 'fulfilled'
+      }
+    }
+
+    return {
+      udpMapped: false,
+      tcpMapped: false
+    }
+  }
+
+  /**
+   * Attaches the listener to TCP and UDP sockets
+   * @param ma address to listen to
+   */
+  async listen(ma: Multiaddr): Promise<void> {
+    if (this.state == State.CLOSED) {
+      throw Error(`Cannot listen after 'close()' has been called`)
+    }
+
+    await this.bind(ma)
+
+    const ownInterface = this.tcpSocket.address() as AddressInfo
+
+    const natSituation = await this.checkNATSituation(ownInterface.address, ownInterface.port)
+
+    log(`NAT situation detected: `, natSituation)
+    const internalInterfaces = getAddrs(ownInterface.port, {
+      useIPv4: true,
+      includePrivateIPv4: true,
+      includeLocalhostIPv4: true
+    })
+
+    if (!natSituation.bidirectionalNAT) {
+      // If any of the interface addresses is the
+      // external address,
+      for (const [index, internalInterface] of internalInterfaces.entries()) {
+        if (internalInterface.address == natSituation.externalAddress) {
+          internalInterfaces.splice(index, 1)
+        }
+      }
+
+      this.addrs.external = [
+        nodeToMultiaddr(
+          {
+            address: natSituation.externalAddress,
+            port: natSituation.externalPort,
+            family: 'IPv4'
+          },
+          this.peerId
+        )
+      ]
+    }
+
+    this.addrs.interface = internalInterfaces.map((internalInterface) =>
+      nodeToMultiaddr(internalInterface, this.peerId)
     )
 
-    // Prevent from sending a STUN request to self
-    let usableStunServers = this.getUsableStunServers(address.port, address.address)
-    await this.determinePublicIpAddress(usableStunServers)
-    await this.updatePublicNodes()
+    this.attachSocketHandlers()
+
+    this._emitListening()
+
+    // Only add relay nodes if node is not directly reachable or running locally
+    if (this.testingOptions.__runningLocally || natSituation.bidirectionalNAT || !natSituation.isExposed) {
+      this.entry.on(RELAY_CHANGED_EVENT, this._emitListening)
```

Comment:
Whenever the list of used relay has changed, emit the `RELAY_CHANGED_EVENT` which reemit the `listening` event to call internal see 

https://github.com/libp2p/js-libp2p/blob/c3700f55d5a0b62182d683ca37258887b24065b9/src/transport-manager.js#L181